Editorial review: Document Wasm bulk memory operations (#44296)
mdn/content: files/en-us/webassembly/reference/data/drop/index.md

---
title: "drop: Wasm data instruction"
short-title: drop
slug: WebAssembly/Reference/Data/drop
page-type: webassembly-instruction
browser-compat: webassembly.data.drop
sidebar: webassemblysidebar
---

> The **`data.drop`** [data instruction](/en-US/docs/WebAssembly/Reference/Data) discards the data contained by a passive `data` definition, freeing up its memory, after the associated `memory` has been initialized via [`memory.init`](/en-US/docs/WebAssembly/Reference/Memory/init).

Reviewer: Same nit as for `elem.drop`: Passive data segments do not have an "associated memory".

```suggestion
The **`data.drop`** [data instruction](/en-US/docs/WebAssembly/Reference/Data) discards the data contained by a passive `data` definition, freeing up its memory, after being used in a [`memory.init`](/en-US/docs/WebAssembly/Reference/Memory/init).
```

Author: Fixed here, and on the `data` summary page.

  )
)
```

```js interactive-example
WebAssembly.instantiateStreaming(fetch("{%wasm-url%}")).then((result) => {
  result.instance.exports.init();
  const memBuffer = result.instance.exports.memory.buffer;
  const memArray = new Uint8Array(memBuffer, 0, 5);
  console.log(new TextDecoder().decode(memArray));
});
```

In the above example, we define a `memory` and a `data` called `$greeting` that contains the string `Hello`. We then invoke `memory.init` to copy the string over to the `memory`. With this done, the `data` is no longer needed, so we call `data.drop` to free up the memory it was using.

## Syntax

```plain
data.drop identifier
```

- `data.drop`
  - : The `data.drop` instruction type. Must always be included first.
- `identifier`
  - : The identifier for the `data` you want to drop. This can be one of the following:
    - `name`
      - : An identifying name [set for the `data`](/en-US/docs/WebAssembly/Reference/Definitions/data#name) when it was first defined. This must begin with a `$` symbol, for example `$my_data`.
    - `index`
      - : The `data`'s index number, for example `0` for the first `data` in the wasm module, `1` for the second, etc.

### Type

```plain

> [] -> []

Reviewer 2: As commented elsewhere, the "Type" section is a new convention for me. What are you trying to say here? Where is the convention introduced and documented? When should it exist and when not exist.

I'm not saying it is wrong, but I don't know how a new reader finding our docs could interpret this. I had the same problem when I first saw the term "TRAP" in Wasm.

Author: AIUI, Type and Trap are well understood by the target audience. A complete beginner to Wasm would definitely not start with these reference docs. 

We are definitely missing some guides to explain this stuff, but we are intending to write those later.

Reviewer: FWIW, I like @hamishwillee's suggestion of titling this section "Operand Stack", but that's for another PR I think.

Reviewer 2: OK, so as per https://github.com/mdn/content/pull/44296#discussion_r3687162056 , agree we should move the structuring of the syntax to another issue/PR. Marking as resolved.
